---
title: "Sanger notebook"
output: html_notebook
---

Sample notebook to process input reads and save both output FASTAs and quality plots. First, set variables for this analysis:

```{r settings}
library(tidyverse)
library(kableExtra)

# What type of sequences are these?
seq_mode <- "16S"
# seq_mode <- "phylotag"

# Where is the vsearch binary?
vsearch_path <- "/Users/bradley.720/miniforge3/envs/culturomics/bin/vsearch"

# Directory containing sequences, and how to rename these
data_dir <- "./seqs/"
text_to_remove <- c("Christian_", "515F_718447_", ".ab1")

# Metadata
gtdb_metadata <- "./data/bac120_metadata.tsv.gz"
gtdbtk_assembly_metadata <- "./data/gtdbtk.bac120.summary.tsv"
if (seq_mode == "16S") {
  database <- "./data/ssu_all.fna"
} else if (seq_mode == "phylotag") {
  database <- "./data/gyr.fna"
}

# Output files
all_sanger_out <- "./concat_sanger_out.fa"
vsearch_outfile <- "./vsearch_output.txt"
final_output_tbl <- "./vsearch_processed_table.tsv"

# PhyloTag specific settings:
#
# Because PhyloTag hits may not be unique *across* samples, we can be explicit about which samples these Sanger sequences came from.
which_samples <- c("FF08034199", "FF08034332", "FF08034525")
# Otherwise:
# which_samples <- NULL
# If your sequences are named in a way that is not helpful, you can use a regex to transform them
phylotag_regex <- c(pattern="^(DS[[:digit:]]+_group_[[:digit:]]+_bin\\.[[:digit:]]+)_.*",
                    reference="\\1")

# Set environment variables for the BASH script section
Sys.setenv(ALL_SANGER_OUT=all_sanger_out)
Sys.setenv(DATABASE=database)
Sys.setenv(VSEARCH_PATH=vsearch_path)
Sys.setenv(SANGER_OUT_DIR=file.path(getwd(), out_dir))
Sys.setenv(VSEARCH_OUT=vsearch_outfile)
```

Now, load and default-trim chromatograms. (If you already have sequences, you can skip to section "vsearch_run".) 

```{r load, echo=TRUE, results='hide', message=FALSE}

source("process_sanger_functions.R")

log <- capture.output(
  trimmed <- load_and_trim(data_dir = data_dir,
                           fn_remove = text_to_remove,
                           TrimmingMethod = "M2"),
  type="message"
)

```


Now, generate chromatograms and show which regions have been auto-trimmed. If you have adjustments based on this, modify the code above, then re-run this chunk.


```{r chromatograms}
makeChromatogram <- function(x, trim5=NULL, trim3=NULL, ...) {
  if (is.null(trim5)) { trim5 <- x@QualityReport@trimmedStartPos }
  if (is.null(trim3)) { trim3 <- x@QualityReport@rawSeqLength - x@QualityReport@trimmedFinishPos }
  sangerseqR::chromatogram(
    x,
    trim5 = trim5,
    trim3 = trim3,
    showtrim=TRUE, ...)
}

for (i in 1:length(trimmed)) {
  x <- trimmed[[i]]
  n <- names(trimmed)[i]
  makeChromatogram(x)
  mtext(n, side=1, adj=1)
}

```


Re-trim any that need manual re-trimming...

```{r manual_trim}

trimmed[["D_D01"]] <- manually_trim(trimmed[["D_D01"]], trim5=100, trim3=121)
makeChromatogram(trimmed[["D_D01"]], width=50)

trimmed[["G_G01"]] <- manually_trim(trimmed[["G_G01"]], trim5=102, trim3=110)
makeChromatogram(trimmed[["G_G01"]])

trimmed[["N_F02"]] <- manually_trim(trimmed[["N_F02"]], trim5=105, trim3=110)
makeChromatogram(trimmed[["N_F02"]])

trimmed[["O_G02"]] <- manually_trim(trimmed[["O_G02"]], trim5=105, trim3=85)
makeChromatogram(trimmed[["O_G02"]])

# This one just looks contaminated
trimmed[["Q_A03"]] <- manually_trim(trimmed[["Q_A03"]], trim5=100, trim3=15)
makeChromatogram(trimmed[["Q_A03"]])

trimmed[["T_D03"]] <- manually_trim(trimmed[["T_D03"]], trim5=120, trim3=120)
makeChromatogram(trimmed[["T_D03"]])

```


You can also re-trim using other methods if you feel it is appropriate (by default, disabled):

```{r adjust}

# To adjust reads based on the output above, modify the below code. Saving this as a new notebook allows you to have a record of how you adjusted trimming parameters. After you run this chunk, be sure to re-run the plots above.

adjustments <- FALSE
if (adjustments) {
  # example
  trim_by_M1 <- c()#c("D_D01", "G_G01", "N_F02", "O_G02", "Q_A03", "T_D03")
  for (n in trim_by_M1) {
    new_trimmed <- sangeranalyseR::updateQualityParam(trimmed[[n]],
                                                       TrimmingMethod = "M1",
                                                       M1TrimmingCutoff = 0.001)
    if ("message" %in% names(new_trimmed)) {
      message(new_trimmed$message)
    } else {
      trimmed[[n]] <- new_trimmed  
    }
  }
  retrim_by_M2_1 <- c("D_D01", "G_G01", "N_F02")
  for (n in retrim_by_M2_1) {
    new_trimmed <- sangeranalyseR::updateQualityParam(trimmed[[n]],
                                                       TrimmingMethod = "M2",
                                                       M2CutoffQualityScore = 18,
                                                       M2SlidingWindowSize = 10)
    if ("message" %in% names(new_trimmed)) {
      message(new_trimmed$message)
    } else {
      trimmed[[n]] <- new_trimmed  
    }
  }
}

#trimmed[["D_D01"]] <- sangeranalyseR::updateQualityParam(trimmed[["D_D01"]], TrimmingMethod="M2", M2CutoffQualityScore = 18, M2SlidingWindowSize = 10)
#makeChromatogram(trimmed[["D_D01"]])

```

Optionally, create quality plots (disabled by default as these take up a lot of memory):

```{r quality}

make_quality_plots <- FALSE
if (make_quality_plots) {
  qc_plots <- htmltools::tagList()
  for (i in 1:length(trimmed)) {
    qc_plots[[i]] <- 
      plotly::as_widget(
        plotly::layout(
          sangeranalyseR::qualityBasePlot(trimmed[[i]]),
          title = list(text=names(trimmed)[i], x=.05)
        )
      )
  }
  names(qc_plots) <- names(trimmed)
  qc_plots
}

```

Finally, write trimmed alignments to disk:

```{r output, echo=TRUE, results='hide', message=FALSE}

out_dir <- "./fasta_output/"
dir.create(out_dir, showWarnings=FALSE)
for (i in 1:length(trimmed)) {
  n <- names(trimmed)[i]
  x <- trimmed[[i]]
  x@readFileName <- paste0(n, ".ab1") # Comment out to preserve the original names
  sangeranalyseR::writeFasta(x, outputDir = out_dir, compress = FALSE)
}

```


Search with vsearch against database:

```{bash vsearch_run}

#These variables is now set above; uncomment if you want to override
#ALL_SANGER_OUT="./concat_sanger_out.fa"
#VSEARCH_OUT="./vsearch_output.txt"
#DATABASE="./data/ssu_all.fna"
#VSEARCH_PATH="/Users/bradley.720/miniforge3/envs/culturomics/bin/vsearch"

# concat
cat $SANGER_OUT_DIR/*.fa > $ALL_SANGER_OUT

# Run with minimum percent ID threshold of 85%
$VSEARCH_PATH --usearch_global $ALL_SANGER_OUT --db $DATABASE --id 0.85 --blast6out $VSEARCH_OUT

```


This shows an alternative running on some phylotag test data:

```{bash vsearch_run_phylotag_test}

#These variables is now set above; uncomment if you want to override
#ALL_SANGER_OUT="./concat_sanger_out.fa"
VSEARCH_OUT="./vsearch_output_phylotag.txt"
DATABASE="./data/gyr.fna"
#VSEARCH_PATH="/Users/bradley.720/miniforge3/envs/culturomics/bin/vsearch"


# Run with minimum percent ID threshold of 85%
$VSEARCH_PATH --usearch_global "./test_sequences.txt" --db $DATABASE --id 0.85 --blast6out $VSEARCH_OUT

# After running this test case, be sure to run in R:
#
# vsearch_output <- "./vsearch_output_phylotag.txt"
# seq_mode <- "phylotag"

```

Next, read in the vsearch output and analyze it:

```{r parse_output}

blast6out_headers <- c("qseqid", "sseqid", "pident", "length", "mismatch","gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore")

vsearch_results <- readr::read_tsv(vsearch_outfile,
                                   col_names=blast6out_headers,
                                   col_types="ccdddddddddd")

if (seq_mode == "16S") {
  if (!("db_metadata" %in% names(.GlobalEnv))) {
    db_metadata <- readr::read_tsv(gtdb_metadata)
  }
} else if (seq_mode == "phylotag") {
  if (!("phylotag_md" %in% names(.GlobalEnv))) {
    phylotag_md <- readr::read_tsv(gtdbtk_assembly_data)
    bin_md <- readr::read_tsv(bin_to_sample) %>% distinct()
    phylo_bin <- left_join(phylotag_md, bin_md, by=c("user_genome"="bin"))
  }
}

# match columns as appropriate to the database
if (seq_mode == "16S") {
  vsearch_results <- vsearch_results %>%
    separate_wider_delim(sseqid, delim="~", names=c("accession","sequence")) %>%
    left_join(., db_metadata) %>%
    separate_taxonomy_with_s(taxa_col = gtdb_taxonomy) %>%
    relocate(qseqid, Family, Species, pident, length)
} else {
  vsearch_results <- vsearch_results %>%
    mutate(user_genome = gsub(phylotag_regex[["pattern"]],
                              phylotag_regex[["reference"]],
                              sseqid)) %>%
    left_join(., phylo_bin) %>%
    separate_taxonomy_with_s(taxa_col = classification) %>%
    relocate(qseqid, Family, Species, sample, user_genome, pident, length)
  if (!is.null(which_samples)) {
    vsearch_results <- filter(vsearch_results, sample %in% which_samples)
  }
}

# whittle down results
vsearch_processed <- vsearch_results %>%
  filter(length >= 50) %>%
  filter(pident >= 90) %>%
  mutate(qseqid = gsub("\\.ab1", "", qseqid)) %>%
  group_by(qseqid) %>%
  slice_max(tibble(pident, length), n=1)
kableExtra::kbl(vsearch_processed) %>% kable_styling()

write_tsv(vsearch_processed, final_output_tbl)
```

